File: pennylane_julia/julia_qubit.py
from pennylane.devices import DefaultQubit
import julia

jl = julia.Julia(compiled_modules=False)

from julia import Main
Main.include('JuliaQubit.jl')

class JuliaQubit(DefaultQubit):
    """ PennyLane Julia deivce.


    Args:
        wires (int): the number of wires to initialize the device with
        shots (int): Currently not applicable
        cache (int): currently not applicable
    """

    name = "Julia Qubit PennyLane plugin"
    short_name = "julia.qubit"
    pennylane_requires = "dev?"
    version = "0.0.1"
    author = "Xanadu Inc. : Christina Lee"

    operations = {
        "PauliX",
        "PauliY",
        "PauliZ",
        "Hadamard",
        "S",
        "T",
        "CNOT",
        "SWAP",
        "CSWAP",
        "Toffoli",
        "CZ",
        "PhaseShift",
        "RX",
        "RY",
        "RZ",
        "Rot",
        "CRX",
        "CRY",
        "CRZ",
        "CRot"
    }

    observables = {"PauliX", "PauliY", "PauliZ", "Hadamard", "Identity"}

    def __init__(self, wires, *, shots=1000, cache=0):
        super().__init__(wires=wires, shots=shots)
        
    @classmethod
    def capabilities(cls):
        capabilities = super().capabilities().copy()
        capabilities.update(
            model="qubit",
            supports_reversible_diff=False,
            supports_inverse_operations=False,
            supports_analytic_computation=True,
            returns_state=True,
        )
        capabilities.pop("passthru_devices", None)
        return capabilities

    def apply(self, operations, rotations=None,**kwargs):
        """ creates self.state and modifies it
        Args:
            operations (list[~.Operation]): operation to apply to the device
        """
        self._state = Main.JuliaQubit.apply_all(self._wires, operations)

        # store the pre-rotated state
        self._pre_rotated_state = self._state

        #Main.JuliaQubit.apply(self._state, )

    #def analytic_probability(self, wires=None):
        """ returns marginal probability

        Args:
            wires (Iterable[Number, str], Number, str, Wires):
            wires to return marginal probabilties for.  Wires not provided are traced out of the system

        Returns:
            List[float]: list of the probabilities
        """
    #   pass

File: pennylane_julia/JuliaQubit.jl
module JuliaQubit2

using PyCall
using TensorOperations

export JuliaQubit

qml = pyimport("pennylane")
devices_mod = pyimport("pennylane.devices")

@pydef mutable struct JuliaQubit <: devices_mod.DefaultQubit

    name = "Julia Qubit PennyLane plugin"
    short_name = "julia.qubit"
    pennylane_requires = "dev?"
    version = "0.0.1"
    author = "Christina Lee"

    operations = [
        "PauliX",
        "PauliY",
        "PauliZ",
        "Hadamard",
        "S",
        "T",
        "CNOT",
        "SWAP",
        "CSWAP",
        "Toffoli",
        "CZ",
        "PhaseShift",
        "RX",
        "RY",
        "RZ",
        "Rot",
        "CRX",
        "CRY",
        "CRZ",
        "CRot"
    ]

    observables = ["PauliX", "PauliY", "PauliZ", "Hadamard", "Identity"]

    function __init__(self, wires, args...; shots=None, cache=0)
            qml.devices.DefaultQubit.__init__(self, wires, shots=shots, cache=cache)
    end

    function apply_operation(self, state, op)
        dev_wires =  [self.wires_map[w] for w in op.wires]
        
        state_indices = -collect(1:self.num_wires)
        op_indices = cat(zeros(Int,length(op.wires)), (1:length(op.wires)), dims=1)

        for (i, w) in enumerate(reverse(dev_wires))
            state_indices[w]= i
            op_indices[i] = -w
        end

        matr = reshape(op.matrix, repeat([2], 2*op.wires.__len__())...)

        return ncon((matr, state), (op_indices, state_indices))
    end

    function apply(self, operations, args...; rotations=None, kwargs...)

        self._state = zeros(Complex{Float64}, repeat([2], self.num_wires)...)
        self._state[1] = 1.0

        self.wires_map = Dict(zip(self._wires, 1:self.num_wires))
        
        for op in operations
            self._state = self.apply_operation(self._state, op)
        end

        self._prerotated_state = self._state
    end

end

end
